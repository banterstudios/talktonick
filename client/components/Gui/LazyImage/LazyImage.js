import React, { Component } from 'react'

import InView from 'client/components/Common/InView'

import PropTypes from 'prop-types'

import { getImage } from 'client/utils/imageUtils'

class LazyImage extends Component {
  static PropTypes = {
    src: PropTypes.string,
    useBgImage: PropTypes.bool
  }

  static defaultProps = {
    useBgImage: false
  }

  constructor (props) {
    super(props)

    this.state = {
      loaded: false,
      error: false,
      isLoading: false,
      visible: null
    }

    this.imagePromise = null
  }

  componentWillUnmount () {
    this.cancelFetchIfFetching()
  }

  cancelFetchIfFetching () {
    if (this.imagePromise && this.imagePromise.cancel) {
      this.imagePromise.cancel()
      this.imagePromise = null
    }
  }

  handleImageLoad = () => {
    this.setState({
      loaded: true,
      isLoading: false,
      error: false
    })
  }

  handleImageError = () => {
    this.setState({
      loaded: false,
      error: true,
      isLoading: false
    })
  }

  fetchImage (src = this.props.src) {
    const { isLoaded, isLoading } = this.state

    if (!src || isLoaded || isLoading) {
      return false
    }

    this.imagePromise = getImage(src)
    .then(this.handleImageLoad)
    .catch(this.handleImageError)
  }

  shouldComponentUpdate (nextProps, { loaded }) {
    return loaded
  }

  componentWillReceiveProps ({ src }, { isVisible }) {
    if (src !== this.props.src && isVisible) {
      this.fetchImage(src)
    }
  }

  handleVisibility = (isVisible) => {
    this.setState({ isVisible })
    isVisible && this.fetchImage()
  }

  renderImage () {
    const {
      useBgImage
    } = this.props
  }

  render () {
    return (
      <InView onChange={this.handleVisibility}>
        { this.renderImage() }
      </InView>
    )
  }
}

export default LazyImage
