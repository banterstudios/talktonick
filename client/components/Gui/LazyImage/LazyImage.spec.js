import React from 'react'
import { mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import LazyImage from './LazyImage'

describe('components/Gui/LazyImage', () => {
  let initialProps

  beforeEach(() => {
    initialProps = {
      src: '',
      alt: '',
      bgPos: '',
      bgSize: '',
      bgRepeat: ''
    }
  })

  describe('@renders', () => {
    it('in default state', () => {
      const component = mount(<LazyImage { ...initialProps } />)

      expect(toJson(component)).toMatchSnapshot()
      component.unmount()
    })

    it('a div with a background image', () => {
      const component = mount(<LazyImage { ...initialProps } useBgImage />)

      expect(toJson(component)).toMatchSnapshot()
      component.unmount()
    })
  })
})
