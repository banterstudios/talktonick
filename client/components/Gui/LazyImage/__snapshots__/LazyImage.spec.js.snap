// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Gui/LazyImage @renders a div with a background image 1`] = `
<LazyImage
  alt=""
  bgPos=""
  bgRepeat=""
  bgSize=""
  src=""
  useBgImage={true}
>
  <InView
    onChange={[Function]}
    resizeCheck={true}
    scrollCheck={true}
    timerCheck={false}
    timerDelay={200}
  >
    <div>
      <glamorous(div)
        loaded={false}
      >
        <div
          className="css-1kyj4b8"
        >
          <glamorous(div)
            bgImage=""
            bgPos=""
            bgRepeat=""
            bgSize=""
          >
            <div
              className="css-r8yd8y"
            />
          </glamorous(div)>
        </div>
      </glamorous(div)>
    </div>
  </InView>
</LazyImage>
`;

exports[`components/Gui/LazyImage @renders in default state 1`] = `
<LazyImage
  alt=""
  bgPos=""
  bgRepeat=""
  bgSize=""
  src=""
  useBgImage={false}
>
  <InView
    onChange={[Function]}
    resizeCheck={true}
    scrollCheck={true}
    timerCheck={false}
    timerDelay={200}
  >
    <div>
      <glamorous(div)
        loaded={false}
      >
        <div
          className="css-1kyj4b8"
        >
          <glamorous(img)
            alt=""
            src=""
          >
            <img
              alt=""
              className="css-1az0zxs"
              src=""
            />
          </glamorous(img)>
        </div>
      </glamorous(div)>
    </div>
  </InView>
</LazyImage>
`;
